src/Service/Kraken/KrakenBuyService.php

```diff
@@ -0,0 +1,107 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Jorijn\Bitcoin\Dca\Service\Kraken;
+
+use Jorijn\Bitcoin\Dca\Client\KrakenClientInterface;
+use Jorijn\Bitcoin\Dca\Exception\KrakenClientException;
+use Jorijn\Bitcoin\Dca\Exception\PendingBuyOrderException;
+use Jorijn\Bitcoin\Dca\Model\CompletedBuyOrder;
+use Jorijn\Bitcoin\Dca\Service\BuyServiceInterface;
+
+class KrakenBuyService implements BuyServiceInterface
+{
+    private const SATOSHIS_IN_A_BITCOIN = '100000000';
+    protected string $lastUserRef;
+    protected KrakenClientInterface $client;
+    protected string $baseCurrency;
+    protected string $tradingPair;
+
+    public function __construct(KrakenClientInterface $client, string $baseCurrency)
+    {
+        $this->client = $client;
+        $this->baseCurrency = $baseCurrency;
+        $this->tradingPair = sprintf('XBT%s', $this->baseCurrency);
+    }
+
+    public function supportsExchange(string $exchange): bool
+    {
+        return 'kraken' === $exchange;
+    }
+
+    public function initiateBuy(int $amount): CompletedBuyOrder
+    {
+        // generate a 32-bit singed integer to track this order
+        $this->lastUserRef = (string) random_int(0, 0x7FFFFFFF);
+
+        $addedOrder = $this->client->queryPrivate('AddOrder', [
+            'pair' => $this->tradingPair,
+            'type' => 'buy',
+            'ordertype' => 'market',
+            'volume' => bcdiv((string) $amount, $this->getCurrentPrice(), 8),
+            'oflags' => 'fciq', // prefer fee in quote currency
+            'userref' => $this->lastUserRef,
+        ]);
+
+        $orderId = $addedOrder['txid'][array_key_first($addedOrder['txid'])];
+
+        // check that its closed
+        $this->checkIfOrderIsFilled($orderId);
+
+        return $this->getCompletedBuyOrder($orderId);
+    }
+
+    public function checkIfOrderIsFilled(string $orderId): CompletedBuyOrder
+    {
+        $trades = $this->client->queryPrivate('OpenOrders', ['userref' => $this->lastUserRef]);
```

Comment:
Use lastUserRef as array form here, use orderId to store associative data. Allows for multiple orders at the same time.

Reply:
Consider using TradesHistory API call here, the order won't show as completed until it is. Might reduce the API load from 2 calls to just 1.